### PR TITLE
(BOLT-44) refactor result handling and implement human output

### PR DESCRIPTION
**Pull request by adreyer**
This commit implements the human output format for bolt and adds a
node_started event to outputters to support this.

It also refactors how results are passed simplifying the number of
result classes. Previously there were two sets of result classes one
that was used in the transport for control flow which was translated to
a result of the second type for display. Now there is only a single set
of result classes which are returned by the transport run_* functions
directly. There are no longer separate classes for success and failure
the success? method should be called to differentiate between the two.
Control flow in transport logic has been refactored to use
raise/rescue/ensure instead of result.then/result.ensure.

TODO:
- In order to maintain compatibility with ExecutionResults the to_h
  method of the result class has not been changed. The to_result method
  should be moved to to_h in the future when ExecutionResult can be
  updated
- The transport methods currently catch exceptions and generate "Result"
  instances from them. This is unecessary since Exceptions are caught in
  the executor as well. The transport methods should raise
  Bolt::Node::BaseError subclasses and let the executor handle them

### CODE REVIEW

File: lib/bolt/cli.rb [hunk 1]
```diff
@@ -5,25 +5,20 @@
...
 
 module Bolt
-  class CLIError < RuntimeError
+  class CLIError < Bolt::Error
     attr_reader :error_code
 
     def initialize(msg, error_code: 1)
-      super(msg)
+      super(msg, "bolt/cli-error")
       @error_code = error_code
```
Comment (MikaelSmith) [inline lib/bolt/cli.rb]: How does error_code get used? Should it be passed as the details argument to `super`?

File: lib/bolt/executor.rb [hunk 2]
```diff
@@ -37,18 +37,18 @@ def on(nodes, callback = nil)
...
             rescue StandardError => ex
-              node.logger.error(ex)
-              Bolt::ExceptionResult.new(ex)
+              Bolt::Result.from_exception(ex)
             ensure
               node.disconnect
             end
           results[node] = result
-          @notifier.notify(callback, node, result) if callback
+          if callback
```
Comment (MikaelSmith) [inline lib/bolt/executor.rb]: Why do we make this optional? I see `callback = nil` in args, but if we never exercised it I'm not sure why we want to preserve that.

File: lib/bolt/node/errors.rb [hunk 3]
```diff
@@ -1,10 +1,10 @@
 module Bolt
   class Node
-    class BaseError < StandardError
+    class BaseError < Bolt::Error
       attr_reader :issue_code
 
       def initialize(message, issue_code)
-        super(message)
+        super(message, kind)
```
Comment (MikaelSmith) [inline lib/bolt/node/errors.rb]: I'd opt for passing `super(message, kind, nil, issue_code)` personally to make it clear that `@issue_code` is handled by the parent. Minor quibble.

File: lib/bolt/node/errors.rb [hunk 4]
```diff
@@ -18,5 +18,22 @@ def kind
...
+    class FileError < BaseError
+      def kind
+        'puppetlabs.tasks/task_file_error'
+      end
+    end
+
+    class EnvironmentVarError < BaseError
+      def initialize(var, val)
+        message = "Could not set environment variable '#{var}' to '#{val}'"
+        super(message, 'ENVVAR_ERROR')
```
Comment (MikaelSmith) [inline lib/bolt/node/errors.rb]: How is this issue code helpful vs `puppetlabs.tasks/environment-var-error`?

File: lib/bolt/node/orch.rb [hunk 5]
```diff
@@ -47,40 +47,29 @@ def _run_task(task, _input_method, arguments)
...
-      # Otherwise create a new result with the captured output
-      result_output = Bolt::Node::ResultOutput.new
-      result_output.stdout << task_result['stdout']
-      result_output.stderr << task_result['stderr']
-      if (task_result['exit_code']).zero?
-        Bolt::Node::Success.new(task_result['stdout'], result_output)
-      else
-        Bolt::Node::Failure.new(task_result['exit_code'], result_output)
-      end
+      Bolt::CommandResult.new(result.value['stdout'], result.value['stderr'], result.value['exit_code'])
```
Comment (MikaelSmith) [inline lib/bolt/node/orch.rb]: We expect to use Bolt::CommandResult for scripts as well?

File: lib/bolt/node/ssh.rb [hunk 6]
```diff
@@ -109,45 +108,55 @@ def execute(command, sudoable: false, **options)
       end
       session_channel.wait
 
-      if status[:exit_code].zero?
+      if result_output.exit_code.zero?
         @logger.debug { "Command returned successfully" }
-        Bolt::Node::Success.new(result_output.stdout.string, result_output)
       else
         @logger.info { "Command failed with exit code #{status[:exit_code]}" }
```
Comment (MikaelSmith) [inline lib/bolt/node/ssh.rb]: `status[:exit_code]` no longer exists, should be `result_output.exit_code`.

File: lib/bolt/node/errors.rb (hunk 4 above)
Comment (MikaelSmith) [reply lib/bolt/node/errors.rb]: I guess every BaseError has an issue code, so that makes sense.

File: lib/bolt/node/ssh.rb [hunk 7]
```diff
@@ -109,45 +108,55 @@ def execute(command, sudoable: false, **options)
...
-      make_tempdir.then do |dir|
-        (yield dir).ensure do
-          execute("rm -rf '#{dir}'")
+      dir = make_tempdir
+      begin
+        yield dir
+      ensure
+        output =  execute("rm -rf '#{dir}'")
+        if output.exit_code != 0
+          logger.warn("Failed to clean up tempdir '#{dir}': #{result.stderr}")
```
Comment (MikaelSmith) [inline lib/bolt/node/ssh.rb]: Where does `result` come from? I think you mean `output`.

File: lib/bolt/node/ssh.rb [hunk 8]
```diff
@@ -167,14 +176,20 @@ def make_wrapper_stringio(task_path, stdin)
...
+
     def with_task_wrapper(remote_task, dir, stdin)
       wrapper = make_wrapper_stringio(remote_task, stdin)
       command = "#{dir}/wrapper.sh"
-      _upload(wrapper, command).then do
-        execute("chmod u+x '#{command}'")
-      end.then do
-        yield command
-      end
+      # TODO: update with_remote script to support this
```
Comment (MikaelSmith) [inline lib/bolt/node/ssh.rb]: I'm not sure what this TODO means.

File: lib/bolt/node/winrm.rb [hunk 9]
```diff
@@ -38,7 +38,7 @@ def disconnect
     end
 
     def shell_init
-      return Bolt::Node::Success.new if @shell_initialized
+      return nil if @shell_initialized
```
Comment (MikaelSmith) [inline lib/bolt/node/winrm.rb]: Should we merge BOLT-72 before this? Seems like they might overlap.

File: lib/bolt/node/winrm.rb [hunk 10]
```diff
@@ -220,12 +221,18 @@ def process_from_extension(path)
     end
 
     def _upload(source, destination)
+      write_remote_file(source, destination)
+      Bolt::Result.new
+    # TODO: we should rely on the executor for this
```
Comment (MikaelSmith) [inline lib/bolt/node/winrm.rb]: When do you plan to address the TODO?

File: lib/bolt/node/winrm.rb [hunk 11]
```diff
@@ -220,12 +221,18 @@ def process_from_extension(path)
...
+    rescue StandardError => ex
+      Bolt::Result.from_exception(ex)
+    end
+
+    def write_remote_file(source, destination)
       @logger.debug { "Uploading #{source} to #{destination}" }
       fs = ::WinRM::FS::FileManager.new(@connection)
+      # TODO: raise FileError here if this fails
```
Comment (MikaelSmith) [inline lib/bolt/node/winrm.rb]: That seems simple, still planning to add it?

File: lib/bolt/node/winrm.rb [hunk 12]
```diff
@@ -236,46 +243,49 @@ def make_tempdir
...
-        dir = value
-        ext = File.extname(file)
-        ext = VALID_EXTENSIONS.include?(ext) ? ext : '.ps1'
-        dest = "#{dir}\\#{File.basename(file, '.*')}#{ext}"
-        Bolt::Node::Success.new
-      end.then do
-        _upload(file, dest)
-      end.then do
+      ext = File.extname(file)
+      ext = VALID_EXTENSIONS.include?(ext) ? ext : '.ps1'
```
Comment (MikaelSmith) [inline lib/bolt/node/winrm.rb]: Ah, this is how we support scripts without an extension on Windows.

File: lib/bolt/node/winrm.rb [hunk 13]
```diff
@@ -236,46 +243,49 @@ def make_tempdir
...
 PS
-        result
       end
     end
 
     def _run_command(command)
-      execute(command)
+      output = execute(command)
+      Bolt::CommandResult.from_output(output)
+    # TODO: we should rely on the executor for this
```
Comment (MikaelSmith) [inline lib/bolt/node/winrm.rb]: Is this another ticket you'll file?

File: lib/bolt/outputter/human.rb [hunk 14]
```diff
@@ -1,27 +1,80 @@
 module Bolt
   class Outputter
     class Human < Bolt::Outputter
+      COLORS = { red: "31",
```
Comment (MikaelSmith) [inline lib/bolt/outputter/human.rb]: Do these get translated on Windows? Or does this only work in *nix?

File: lib/bolt/outputter/json.rb [hunk 15]
```diff
@@ -15,10 +15,17 @@ def print_head
...
+        when :node_result
+          print_result(node, event[:result])
+        end
+      end
+
       def print_result(node, result)
         item = {
           name: node.uri,
-          status: result.is_a?(Bolt::ErrorResult) ? 'failure' : 'success',
+          status: result.success? ? 'failure' : 'success',
```
Comment (MikaelSmith) [inline lib/bolt/outputter/json.rb]: This seems reversed... if `success?` is true, wouldn't we use `success`?

File: lib/bolt/result.rb [hunk 16]
```diff
@@ -1,69 +1,71 @@
...
 
     def to_result
-      {
-        '_error' => {
-          'issue_code' => @issue_code,
-          'kind' => @kind,
-          'msg' => @message
-        }
-      }
+      # TODO: This should be to_h but we need to update the plan functions to
```
Comment (MikaelSmith) [inline lib/bolt/result.rb]: Is there a PUP ticket related to that? Or a BOLT ticket (now that we moved run_* to Bolt)?

**Comment by MikaelSmith** [comment]
I didn't look at tests yet, going to leave that for after initial feedback is addressed.

File: lib/bolt/cli.rb (hunk 1 above)
Comment (adreyer) [reply lib/bolt/cli.rb]: It's the exit code for bolt

File: lib/bolt/node/winrm.rb (hunk 9 above)
Comment (Iristyle) [reply lib/bolt/node/winrm.rb]: Yes, please merge #153 first

File: lib/bolt/cli.rb (hunk 1 above)
Comment (MikaelSmith) [reply lib/bolt/cli.rb]: Ah, ok.

File: lib/bolt/executor.rb (hunk 2 above)
Comment (adreyer) [reply lib/bolt/executor.rb]: That is how the boltlib functions call it
https://github.com/puppetlabs/bolt/blob/master/modules/boltlib/lib/puppet/functions/run_task.rb#L76

File: lib/bolt/node/errors.rb (hunk 4 above)
Comment (adreyer) [reply lib/bolt/node/errors.rb]: issue code is to help with reverse translation in the future. It should match the message template.

File: lib/bolt/node/orch.rb (hunk 5 above)
Comment (adreyer) [reply lib/bolt/node/orch.rb]: Yeah I considered renaming it. Any suggestions?

Comment (MikaelSmith) [reply lib/bolt/node/orch.rb]: It seems ok, just surprised me initially.

File: lib/bolt/node/ssh.rb (hunk 8 above)
Comment (adreyer) [reply lib/bolt/node/ssh.rb]: There are too many similar functions here. We should be able to refactor them into fewer more like the winrm transport.

File: lib/bolt/node/winrm.rb (hunk 9 above)
Comment (adreyer) [reply lib/bolt/node/winrm.rb]: It looks like you did that?

File: lib/bolt/node/winrm.rb (hunk 10 above)
Comment (adreyer) [reply lib/bolt/node/winrm.rb]: As part of the transport stabilization

File: lib/bolt/node/winrm.rb (hunk 11 above)
Comment (adreyer) [reply lib/bolt/node/winrm.rb]: I'm not sure what the failure cases look like or what message to generate

File: lib/bolt/node/winrm.rb (hunk 9 above)
Comment (MikaelSmith) [reply lib/bolt/node/winrm.rb]: Yes, done.

File: lib/bolt/outputter/human.rb (hunk 14 above)
Comment (adreyer) [reply lib/bolt/outputter/human.rb]: presumably the latter

File: lib/bolt/result.rb (hunk 16 above)
Comment (adreyer) [reply lib/bolt/result.rb]: unfortunately ExecutionResult has a from_bolt method that has to change.

Comment (MikaelSmith) [reply lib/bolt/result.rb]: That's do-able, just curious what the plan was to do it.

File: lib/bolt/outputter/human.rb [hunk 17]
```diff
@@ -1,27 +1,80 @@
...
+        end
+
+        if result.message
+          @stream.puts(remove_trail(indent(2, result.message)))
+        end
+
+        if result.instance_of? Bolt::TaskResult
+          @stream.puts(indent(2, ::JSON.pretty_generate(result.value)))
+        elsif result.instance_of? Bolt::CommandResult
+          @stream.puts(indent(2, "STDOUT:"))
```
Comment (MikaelSmith) [inline lib/bolt/outputter/human.rb]: Should thees be printed even if stdout or stderr are empty?

Comment (adreyer) [reply lib/bolt/outputter/human.rb]: I think it's useful to at least have the stdout key printed to make it clear it was empty.

**Pull request merged**